numpy/ma/core.py

```diff
@@ -3118,6 +3135,84 @@ def __array_wrap__(self, obj, context=None):
 
         return result
 
+    def __array_ufunc__(self, np_ufunc, method, *inputs, **kwargs):
+        """
+        MaskedArray capability for ufuncs
+
+        Handle masked versions of ufuncs if they are implemented within
+        the MaskedArray module. If the masked ufunc is not implemented,
+        this falls back to the standard numpy ndarray ufunc, which we
+        then call with the ndarray view of the input data.
+
+        """
+        # Get the equivalent masked version of the numpy function
+        # if it is in the module level functions
+        ma_ufunc = np.ma.__dict__.get(np_ufunc.__name__, np_ufunc)
+        if ma_ufunc is np_ufunc:
+            # We didn't have a Masked version of the ufunc, so we need to
+            # call the ndarray version to prevent infinite recursion.
+            return super().__array_ufunc__(np_ufunc, method, *inputs, **kwargs)
```

Comment:
You will need to unwrap the data arrays here and then call things again.  On the plus side, it may actually make sense to always just do that.
Also, you should not use `super()` here, but rather call the original ufunc I think, because after unwrapping it could be that you need to dispatch to a different argument.

There is the small change that the ufuncs behave a bit different from the `np.ma.ufunc` version (and maybe even better?): see [gh-22347](https://github.com/numpy/numpy/issues/22347#issuecomment-1268702905).
So there might even be a point of just ignoring what `np.ma.func` does.

After the fallout from the "simple" gh-22046, I am a bit hesitant about changing these subtleties, we probably need to test very carefully and I am wondering if we should push it of for a "2.0" (which I hope to bring up now for in a year).  In that case, a possible approach could be to hide these definitions behind an environment variable (feature flag) maybe.